### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by lynn**
As mentioned in my email, this makes `array string *` and `array.to_gs` not quadratic.

I tested the resulting change on 1906 GolfScript solutions that Sisyphus in the code.golf Discord had scraped from http://golf.shinh.org. I wondered if I could make this test suite run a little faster so I used ruby-prof to benchmark the execution…

It turns out that a lot of time is spent on `Gint.new` and `Gint.val`. So I had the idea to get rid of `Gint` and instead work with plain `Numeric` everywhere. This works and makes all programs about 25% faster, but this PR much more complicated 🙂 

Feeling confident about having an exhaustive test suite I made some other performance improvements here and there.

BTW, to my surprise, the interpreter is supported just fine by Ruby 3 even though it was written for 1.9! I only have to run `ruby --encoding ASCII-8BIT golfscript.rb` so that it doesn't complain about non-UTF-8 strings.

### CODE REVIEW

File: golfscript.rb [hunk 1]
```diff
@@ -420,10 +332,8 @@ def compile(tokens=scan(/[a-zA-Z_][a-zA-Z0-9_]*|'(?:\\.|[^'])*'?|"(?:\\.|[^"])*"
 	end
 end
 def gpop
-	($lb.size-1).downto(0){|i|
-		break if $lb[i]<$stack.size
-		$lb[i]-=1
-	}
+	i=$lb.size
+	$lb[i] -= 1 while i>0 && $lb[i-=1] >= $stack.size
 	$stack.pop
```
Comment (lynn) [inline golfscript.rb:337]: `gpop` is called super often, so I tried rewriting it a couple of ways to squeeze more performance out (honestly mostly for fun at this point). It seems this is a bit faster than a block.

File: golfscript.rb [hunk 2]
```diff
@@ -2,153 +2,116 @@
...
+		bs << 93
+		Gstring.new(bs)
 	end
+	def go; $stack<<self; end
 	def class_id; 1; end
-	def coerce(b)
-		if b.class == Gint
-			b.coerce(self).reverse
-		elsif b.class == Gstring
+	def gscoerce(b)
```
Comment (lynn) [inline golfscript.rb:110]: I had to rename `coerce` to `gscoerce` because there is already a method named `coerce` on Ruby's `Numeric` and it was messing things up.

File: golfscript.rb [hunk 3]
```diff
@@ -162,23 +125,23 @@ def rightparen
...
 			return b*self if self.class == Gstring && b.class == Garray
-			return self/Gint.new(1)*b if self.class == Gstring
+			return Garray.new(@val.map{|n|Gstring.new([n])})*b if self.class == Gstring
 			return b.factory([]) if @val.size<1
-			r=@val.first
-			r,x=r.coerce(b) if r.class != b.class #for size 1
-			@val[1..-1].each{|i|r=r+b+i}
+			r=@val.first.dup
+			r,x=r.gscoerce(b) if r.class != b.class #for size 1
+			@val[1..-1].each{|i|r=r.concat(b); r=r.concat(i)}
```
Comment (lynn) [inline golfscript.rb:136]: This is the quadraticity fix for `*`

File: golfscript.rb [hunk 4]
```diff
@@ -2,153 +2,116 @@
...
-# 			when Gint then [i]
-# 			when Garray then i.flatten.val
-# 			when Gblock then i.val
-# 			end
-# 		})
-# 	end
-		#use Peter Taylor's fix to avoid quadratic flatten times
+		r = []
+		@val.each {|i| r.concat(i.to_gs.val) }
+		Gstring.new(r)
```
Comment (lynn) [inline golfscript.rb:87]: This is the quadraticity fix for `Garray#to_gs`

File: golfscript.rb [hunk 5]
```diff
@@ -257,79 +195,57 @@ class Gstring < Garray
 	def initialize(a)
 		@val=case a
 			when NilClass then []
-			when String then a.unpack('C*').map{|i|Gint.new(i)}
+			when String then a.bytes
```
Comment (darrenks) [inline golfscript.rb]: Hey, I believe this was not in ruby 1.8. It may be turn out to be unreasonable to try to have code compatible with 1.8 - 3.0+ but I'd like to try if possible.

**Review by darrenks — CHANGES_REQUESTED** [review]
Hey Lynn thanks for the major quadratic improvements and general speed improvements. Would it be possible for you to send me the code you used to test it on the ~1000 programs? That would be very useful for me in reviewing future changes and I really appreciate the thoroughness.

Only issue I see far is the note about 1.8.

Also I will be traveling for ~1.5 weeks and might not be able to fully review until back from that, so please be patient.

**Comment by lynn** [comment]
I added the test suite code. When you first run `python3 test_suite.py` it tells you what file to download and where to extract it 👍 

image

File: test_suite.py [hunk 6]
```diff
@@ -12,11 +12,15 @@
...
         sys.exit(1)
+
+    passes = fails = 0
     for prob in sorted(problems, key=lambda x: x.name):
         # Luck / weird problems
-        if prob.name in ["123", "123 Reloaded", "Error", "Helloworld", "Inverse Quine", "Not Quine", "Quine", "Timeout"]: continue
-        # Ruby version stuff
+        if prob.name in ["123", "123 Reloaded", "Error", "Inverse Quine", "Not Quine", "Palindromic Quine", "Quine", "Timeout"]: continue
+        # Ruby version / float / rational trickery
         if prob.name in ["area of triangle", "0_5 broken keyboard", "Cancel fractions", "Equal Temperament", "MIDI note number to frequency", "Numloop"]: continue
```
Comment (darrenks) [inline test_suite.py]: "Concentric Circles" too

**Comment by darrenks** [comment]
I've read through all the changes and tried to understand them. I believe I'm ok with everything except the change from Gint to using Numeric. It complicates the code and makes it harder to reason about (not that the existing Golfscript code is good in that regard). I would guess that you have covered all your tracks and that it wouldn't cause any bugs... But I'm not 100% sure, since these tests are only features that have been useful in actual codegolf problems rather than all edge cases possible. And Golfscript lacks any unit tests. I considered trying to write Golfscript like that originally (for the other types too) but got too confused :)

I'm definitely for asymptotic performance improvements or even simple constant improvements (of which you made several), but I think if we really wanted the language to be fast we'd need to start over from scratch and build a JIT or an interpreter in a fast language, but this would be difficult due to the extreme dynamic nature of the language (I'm referring to rarely used features like overwriting builtins dynamically).

Also I'm a bit of a noob when it comes to managing pull requests in github so bear with me if you change this initial commit that the others are stacked on. We will get the other changes in though!

Thanks for your help making gs faster.

File: test_suite.py [hunk 7]
```diff
@@ -12,11 +12,15 @@
...
+        # Sisyphus's dump was unluckily case-insensitive, so these have wrong answers mixed in:
+        if prob.name.lower() in ["helloworld", "christmas tree", "multiplication table"]: continue
```
Comment (darrenks) [inline test_suite.py:25]: Also on my system (which is case insensitive), "Tic Tac Toe" is broken due to this issue
